java/client/src/main/java/glide/connectors/handlers/PubSubMessageQueue.java

```diff
@@ -0,0 +1,82 @@
+/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
+package glide.connectors.handlers;
+
+import glide.api.models.PubSubMessage;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.Future;
+
+/** A FIFO message queue for {@link PubSubMessage} backed by {@link ConcurrentLinkedDeque}. */
+public class PubSubMessageQueue {
+    /** The queue itself. */
+    private final ConcurrentLinkedDeque<PubSubMessage> mq = new ConcurrentLinkedDeque<>();
+
+    /**
+     * The head of the queue stored aside as a {@link Future}. Stores a promise to the first message.
+     */
+    private CompletableFuture<PubSubMessage> head = new CompletableFuture<>();
+
+    /** An object to synchronize threads. */
+    private final Object lock = 42;
```

Comment:
Usual best practice is to lock on new Object(). This can actually expose your lock btw cuz you might get the Integer 42 from a pool.

Reply:
Probably I can lock on `state`. Isn't this better?

Reply:
You actually don't want to do this:
1. The object referenced by state keeps mutating, so the object you synchronize on keeps changing, meaning your synchronized blocks are not locking the same thing potentially.
2. An enum entry is basically a singleton, so if you lock on an enum entry, unrelated instances of the PubSubMessageQueue can lock on the same object, reducing parallelism needlessly.

The ideal object to use as a lock for synchronized is a private final newly-created variable. Note that "this" isn't a good choice because external callers can synchronize on your object.